File: Data Structures & Algorithms/house-robber/submission-1.java
class Solution {
    public int rob(int[] nums) {
        int[] money = new int[nums.length];
        int n = nums.length;

        if(n == 1)
            return nums[0];

        money[0] = nums[0];
        money[1] = nums[1];

        for(int i = 2; i < n; i++){
            money[i] = Math.max(nums[i] + money[i-2], nums[i] + (i >= 3 ? money[i-3] : 0));
        }

        return Math.max(money[n-1], money[n-2]);
    }
}

File: Data Structures & Algorithms/house-robber-ii/submission-1.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        def rob_case(houses : List[int]) -> int:
            skip_current_house, rob_current_house = 0, 0

            # skip_current_house -> get max that we have seen till now -> max of skipping last house and robbing last house
            # rob_current_house -> skip last house and get this house
            for house in houses:
                skip_current_house, rob_current_house = max(skip_current_house, rob_current_house), skip_current_house + house
            return max(skip_current_house, rob_current_house)
        
        if len(nums) == 1:
            return nums[0]
            
        return max(rob_case(nums[1:]), rob_case(nums[0 : -1]))